File: leetcode/020.valid-parentheses.easy/solution.js
/**
 * @param {string} s
 * @return {boolean}
 */
const isValid = (s) => {
  const stack = []
  const characters = ['(', '{', '[']
  const characterMap = {
    '}': '{',
    ']': '[',
    ')': '(',
  }
  for (let i = 0; i < s.length; i += 1) {
    let char = s.charAt(i)
    if (characters.indexOf(char) > -1) {
      stack.push(char)
    } else if (characterMap[char]) {
      if (stack.slice(-1)[0] === characterMap[char]) {
        stack.pop()
      } else {
        return false
      }
    } else {
      return false
    }
  }
  return !stack.length
}

File: leetcode/020.valid-parentheses.easy/solution.py
class Solution:
    def isValid(self, s: str) -> bool:
        map = {
            ']': '[',
            ')': '(',
            '}': '{'
        }
        arr = []
        for char in s:
            if char not in map:
                arr.append(char)
            elif len(arr) and arr.pop() == map[char]:
                continue
            else:
                return False
        return len(arr) == 0

if __name__ == "__main__":
    # result = Solution().isValid('()')
    result = Solution().isValid('(())')

    print(result)
